text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+ - `std::mem::align_of::<&mut (&str,())>`: `_RN12std_a1b2c3d43mem8align_ofVIWTRrvEEE`
+
+There's one more thing we have to take into account for generic functions: The compiler may produce "crate-local" copies of a monomorphization. That is, if there is a function `foo<T>` which gets used as `foo<u32>` in two different crates, the compiler (depending on the optimization level) might generate two distinct functions at the LLVM IR level, each with it's own symbol. In order to support this without running into conflicts, symbol names for monomorphizations must include the id of the crate they are instantiated for. This scheme does this by appending an `<crate-id>` suffix to the symbol. So for example the mangling for `std::mem::align_of::<usize>` would actually look like this:
+
+```
+_RN12std_a1b2c3d43mem8align_ofVIjEE12foo_a1b2c3d4 (for crate "foo[a1b2c3d4])
+_RN12std_a1b2c3d43mem8align_ofVIjEE12bar_11223344 (for crate "bar[11223344])
+```
+
+
+### Closures and Closure Environments
+
+The scheme needs to be able to generate symbol names for the function containing the code of a closure and it needs to be able to refer to the type of a closure if it occurs as a type argument. As closures don't have a name, we need to generate one. The scheme proposes to use the namespace and disambiguation mechanisms already introduced above for this purpose. Closures get their own "namespace" (i.e. they are neither in the type nor the value namespace), and each closure has an empty name with a disambiguation index (like for macro hygiene) identifying them within their parent. The full name of a closure is then constructed like for any other named item:
+
+```rust
+mod foo {
+  fn bar(x: u32) {
+    let a = |x| { x + 1 }; // local name: 0C
+    let b = |x| { x + 2 }; // local name: 0Cs_
+
+    a(b(x))
+  }
+}
+
+```
+
+In the above example we have two closures, the one assigned to `a` and the one assigned to `b`. The first one would get the local name `0C` and the second one the name `0Cs_`. The `0` signifies the length of their (empty) name. The `C` is the namespace tag, analogous to the `V` tag for the value namespace. The `s_` for the second closure is the disambiguation index (index `0` is, again, encoded by not appending a suffix). Their full names would then be `N15mycrate_4a3b56d3foo3barV0CE` and `N15mycrate_4a3b56d3foo3barV0Cs_E` respectively.
+
+
+### Methods
+
+Methods are nested within `impl` or `trait` items. As such it would be possible to construct their symbol names as paths like `my_crate::foo::{{impl}}::some_method` where `{{impl}}` somehow identifies the the `impl` in question. Since `impl`s don't have names, we'd have to use an indexing scheme like the one used for closures (and indeed, this is what the compiler does internally). Adding in generic arguments to, this would lead to symbol names looking like `my_crate::foo::impl'17::<u32, char>::some_method`.
```

Comment:
In the interest of keeping this RFC sufficiently detached from current implementation details, can we use some more general placeholder notation, such as `<impl>`, instead of `{{impl}}`?

Reply:
It's just an example of how not to do it. The `{{xyz}}` notation is meant to remind of what some templating engines use, not what the compiler did at some point. But I can change it to `<impl>` if you prefer that.